src/smt/optimization_solver.cpp

```diff
@@ -46,92 +46,371 @@ OptResult OptimizationSolver::checkOpt()
 
-    // We need to save the value since we need the model value just before the
-    // unsat call
-    Assert(!value.isNull());
-    d_savedValue = value;
+  // value equals to upperBound
+  value = optChecker->getValue(target);
+
+  // this gets the bitvector!
+  BitVector bvValue = value.getConst<BitVector>();
+  unsigned int bvSize = bvValue.getSize();
 
-    // increment on the model-value of objective:
-    // if we're maximizing increment = objective > old_objective value
-    // if we're minimizing increment = objective < old_objective value
-    if (d_activatedObjective.getType() == OBJECTIVE_MAXIMIZE)
+  // lowerbound
+  BitVector lowerBound = ((this->d_isSigned) ? (BitVector::mkMinSigned(bvSize))
+                                             : (BitVector::mkZero(bvSize)));
+  // upperbound must be a satisfying value
+  // and value == upperbound
+  BitVector upperBound = bvValue;
+
+  Kind LTOperator =
+      ((d_isSigned) ? (kind::BITVECTOR_SLT) : (kind::BITVECTOR_ULT));
+  Kind GEOperator =
+      ((d_isSigned) ? (kind::BITVECTOR_SGE) : (kind::BITVECTOR_UGE));
+  BitVector pivot;
+  while (true)
+  {
+    if (d_isSigned)
     {
-      increment = nm->mkNode(kind::GT, d_activatedObjective.getNode(), value);
+      if (!lowerBound.signedLessThan(upperBound)) break;
     }
     else
     {
-      increment = nm->mkNode(kind::LT, d_activatedObjective.getNode(), value);
+      if (!lowerBound.unsignedLessThan(upperBound)) break;
     }
-    optChecker->assertFormula(increment);
-    loop_r = optChecker->checkSat();
+    pivot = computeAverage(lowerBound, upperBound, d_isSigned);
+    optChecker->push();
+    // lowerBound <= target < pivot
+    optChecker->assertFormula(
+        nm->mkNode(kind::AND,
+                   nm->mkNode(GEOperator, target, nm->mkConst(lowerBound)),
+                   nm->mkNode(LTOperator, target, nm->mkConst(pivot))));
+    intermediateSatResult = optChecker->checkSat();
+    if (intermediateSatResult.isUnknown() || intermediateSatResult.isNull())
+    {
+      return std::make_pair(OptResult::OPT_UNKNOWN, value);
+    }
+    if (intermediateSatResult.isSat() == Result::SAT)
+    {
+      value = optChecker->getValue(target);
+      upperBound = value.getConst<BitVector>();
+    }
+    else if (intermediateSatResult.isSat() == Result::UNSAT)
+    {
+      if (lowerBound == pivot)
+      {
+        // lowerBound == pivot ==> upperbound = lowerbound + 1
+        // and lowerbound <= target < upperbound is UNSAT
+        // return the upperbound
+        return std::make_pair(OptResult::OPT_OPTIMAL, value);
+      }
+      else
+      {
+        lowerBound = pivot;
+      }
+    }
+    else
+    {
+      return std::make_pair(OptResult::OPT_UNKNOWN, value);
+    }
+    optChecker->pop();
   }
-
-  return OPT_OPTIMAL;
+  return std::make_pair(OptResult::OPT_OPTIMAL, value);
 }
 
-void OptimizationSolver::activateObj(const Node& obj, const int& type)
+std::pair<OptResult, Node> OMTOptimizerBitVector::maximize(
```

Comment:
Perhaps there is enough shared code with `minimize` to create a shared helper function to be used by both?

Reply:
I would like to say fusing max/min into one makes it very messy. fusing signed/unsigned into one function is already messy, and it's even harder to review the algorithm if you try to fuse 4 cases into 1 function.